homeassistant/components/xiaomi_miio/__init__.py

```diff
@@ -108,6 +115,48 @@ def get_platforms(config_entry):
     return []
 
 
+def _async_update_data_default(hass, device):
+    async def update():
+        """Fetch data from the device using async_add_executor_job."""
+        try:
+            async with async_timeout.timeout(10):
+                state = await hass.async_add_executor_job(device.status)
+                _LOGGER.debug("Got new state: %s", state)
+                return state
+
+        except DeviceException as ex:
+            raise UpdateFailed(ex) from ex
+
+    return update
+
+
+def _async_update_data_vacuum(hass, device: Vacuum):
+    async def update():
+        """Fetch data from the device using async_add_executor_job."""
+        try:
+            async with async_timeout.timeout(10):
+                results = await asyncio.gather(
+                    hass.async_add_executor_job(device.status),
+                    hass.async_add_executor_job(device.dnd_status),
+                    hass.async_add_executor_job(device.last_clean_details),
+                    hass.async_add_executor_job(device.consumable_status),
+                    hass.async_add_executor_job(device.clean_history),
+                )
```

Comment:
Yes, we should create a function that does all these sync calls and then schedule that function once on the executor here.

Reply:
Made them sequential. @rytilahti when you have a chance, pls let me know if it solved the issue for you. Looks good on my end.

Reply:
It is indeed much better! What should be changed is to modify the vacuum platform to use the data from the coordinator instead of doing separate polling (which causes currently failures now and then).

Reply:
aah good suggestion, will look into this as well then.